### PR TITLE
table cell selection improvement

### LINKED ISSUE #2727: Cannot select multiple cells by holding Shift key
- Azure Data Studio Version: 1.0.0

Steps to Reproduce:

1. Run any query that returns multiple columns or rows.
2. Click on a cell of the `Query Results`
3. Hold <kbd>Shift</kbd> and click on another cell

Expected Behavior:
The cells in the range of the first and the last cell clicked should be selected

Actual Behavior:
Only the last cell clicked is selected

### PR DESCRIPTION
**Pull request by alanrenmsft**
Fixes the following issues:
1. This PR fixes #2727
2. Support cell selections on macOS (previously the multi-selection using keyboard never worked)
3. The rowNumber plugin and cellSelectionModel is conflicting since they both do selection management; it is just a matter of the event registration order to decide who gets called first. To fix the issue, I am adding options to the plugins and the consumer of the components can set those options properly. 

**NOTE**: The multi-selection key is CTRL for Windows/Linux, CMD key for macOS.

**Scenarios**
1. Select all - click the header of row number column
select-all

4. Select a single column - click the column header
select-single-column

5. Select columns individually - click the header of the columns with CTRL/CMD key pressed
select-multiple-columns-ctrl

6. Select columns in a range - select the start column (a cell or entire column), then select the end column header with SHIFT key pressed
select-multiple-columns-shift

7. Select a row - click the row number cell
single-row-selection

8. Select muliple rows individually - click the row number cells with CTRL/CMD key pressed
multi-row-selection-ctrl

9. Select rows in a range - select the start row (a cell or the row header), then select the end row header with SHIFT key pressed
multi-row-selection-shift

10. Select cells in a range with mouse - press and hold the select button in the start cell, then move the cursor to the end cell and release the button
cell-range-selection-mouse

11. Select cells in a range with keyboard only - select the start cell, press and hold SHIFT key, then use the arrow keys to select the range
cell-range-selection-keyboard-only

12. Select cells in a range with mouse and SHIFT key - select the start cell, press and hold SHIFT key, then click the end cell
cell-range-selection-mouse-shift

13.  Select multiple ranges - Select a range with the options listed above and then with CTRL/CMD pressed and keep adding new selections.
select-multiple-ranges

### CODE REVIEW

File: src/sql/base/browser/ui/table/plugins/rowNumberColumn.plugin.ts [hunk 1]
```diff
@@ -6,15 +6,19 @@
 import { FilterableColumn } from 'sql/base/browser/ui/table/interfaces';
 
 export interface IRowNumberColumnOptions {
-	numberOfRows: number;
```
Comment (alanrenmsft) [inline src/sql/base/browser/ui/table/plugins/rowNumberColumn.plugin.ts:9]: this thing is unused and also doesn't make sense, so removing it.

File: src/sql/base/browser/ui/table/plugins/rowNumberColumn.plugin.ts [hunk 2]
```diff
@@ -29,7 +33,7 @@ export class RowNumberColumn<T> implements Slick.Plugin<T> {
 	}
 
 	private handleClick(e: MouseEvent, args: Slick.OnClickEventArgs<T>): void {
-		if (this.grid.getColumns()[args.cell].id === 'rowNumber') {
+		if (this.grid.getColumns()[args.cell].id === 'rowNumber' && this.options.autoCellSelection !== false) {
```
Comment (aasimkhan30) [inline src/sql/base/browser/ui/table/plugins/rowNumberColumn.plugin.ts:36]: Why not this.options.autoCellSelection

File: src/sql/base/browser/ui/table/plugins/rowNumberColumn.plugin.ts [hunk 3]
```diff
@@ -38,7 +42,7 @@ export class RowNumberColumn<T> implements Slick.Plugin<T> {
 	}
 
 	private handleHeaderClick(e: MouseEvent, args: Slick.OnHeaderClickEventArgs<T>): void {
-		if (args.column.id === 'rowNumber') {
+		if (args.column.id === 'rowNumber' && this.options.autoCellSelection !== false) {
```
Comment (aasimkhan30) [inline src/sql/base/browser/ui/table/plugins/rowNumberColumn.plugin.ts:45]: Same here

File: src/sql/base/browser/ui/table/plugins/rowNumberColumn.plugin.ts (hunk 2 above)
Comment (alanrenmsft) [reply src/sql/base/browser/ui/table/plugins/rowNumberColumn.plugin.ts:36]: the default value for this option is true

**Review by aasimkhan30 — APPROVED** [review]

**Comment by chlafreniere** [comment]
I'm positive that you looked at this, but what kind of accessibility testing did you do with this change?

File: src/sql/base/browser/ui/table/plugins/cellSelectionModel.plugin.ts [hunk 4]
```diff
@@ -228,29 +239,33 @@ export class CellSelectionModel<T> implements Slick.SelectionModel<T, Array<Slic
...
-		this.grid.setActiveCell(selectedRange.toRow, selectedRange.toCell);
-		this.setSelectedRanges(ranges);
+		// When the CTRL key is pressed, we need to merge the new selection with the existing selections.
+		const rangesToBeMerged: Slick.Range[] = this.isMultiSelection(e) ? this.getSelectedRanges() : [];
+		const result = this.insertIntoSelections(rangesToBeMerged, newlySelectedRange);
+		this.setSelectedRanges(result);
 
-		e.preventDefault();
-		e.stopImmediatePropagation();
+		// Find out the new active cell
```
Comment (Charles-Gagnon) [inline src/sql/base/browser/ui/table/plugins/cellSelectionModel.plugin.ts:266]: Would be good to explain this logic in the comment too

File: src/sql/base/browser/ui/table/plugins/rowNumberColumn.plugin.ts (hunk 2 above)
Comment (Charles-Gagnon) [reply src/sql/base/browser/ui/table/plugins/rowNumberColumn.plugin.ts:36]: I believe he's asking why the check isn't `(this.grid.getColumns()[args.cell].id === 'rowNumber' && this.options.autoCellSelection)`

Which the reason is that the "default" is supposed to be true, but if no options are given then options defaults to an empty object. 

@alanrenmsft I think it'd be clearer to make a default options object and then merge the two in the constructor so you're sure that it actually has a value (and the value is true) if nothing is specified.

**Review by Charles-Gagnon — APPROVED** [review]
Definitely should set up a bug bash for this at some point (and make sure it's called out for extra testing during signoff).

File: src/sql/base/browser/ui/table/plugins/rowNumberColumn.plugin.ts (hunk 2 above)
Comment (alanrenmsft) [reply src/sql/base/browser/ui/table/plugins/rowNumberColumn.plugin.ts:36]: yes, that's what I am working on. mixin the options with defaults.

File: src/sql/base/browser/ui/table/plugins/cellSelectionModel.plugin.ts (hunk 4 above)
Comment (alanrenmsft) [reply src/sql/base/browser/ui/table/plugins/cellSelectionModel.plugin.ts:268]: done

**Comment by alanrenmsft** [comment]
@chlafreniere , there are quite a few issues with the grid, I have a master issue tracking all them: https://github.com/microsoft/azuredatastudio/issues/20855.

**Pull request merged**